### PR TITLE
feat: read json lazily to tolerate incorrect closing brackets

### LINKED ISSUE #32: Tune assistant for anthropic models

### PR DESCRIPTION
**Pull request by Oleksii-Klimov**

### CODE REVIEW

File: aidial_assistant/application/prompts.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def build(self, **kwargs) -> Template:
     {
       "command": "<command name>",
       "args": [
-        // <array of arguments>
+        "arr1", "arg2", ...
```
Comment (adubovik) [inline aidial_assistant/application/prompts.py]: Typo

Comment (Oleksii-Klimov) [reply aidial_assistant/application/prompts.py]: Fixed.

File: aidial_assistant/json_stream/json_parser.py [hunk 2]
```diff
@@ -38,34 +46,43 @@ def string_node(node: JsonNode) -> JsonString:
...
-                task = tg.create_task(JsonParser._parse_root(root, stream))
-                try:
-                    yield await root.node()
-                finally:
-                    await task
-        except ExceptionGroup as e:
-            raise e.exceptions[0]
+            normalised_stream = JsonNormalizer(stream)
+            char = await normalised_stream.apeek()
+            if char == JsonObject.token():
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_parser.py]: Maybe introduce the same interface for all these classes, so that dispatching looks uniform:

```python
class RootNodeResolver(NodeResolver):
    @override
    async def resolve(self, stream: CharacterStream) -> JsonNode:
        try:
            normalised_stream = JsonNormalizer(stream)
            char = await normalised_stream.apeek()
            if JsonObject.is_token():
                return JsonObject.parse(stream, self)

            if JsonArray.is_token():
                return JsonArray.parse(stream, self)

            if JsonString.is_token():
                return JsonString.parse(stream)

            if JsonNumber.is_token(char):
                return JsonNumber.parse(stream)

            if JsonNull.is_token(char):
                return JsonNull.parse(stream)

            if JsonBoolean.is_token(char):
                return JsonBoolean.parse(stream)
        except StopAsyncIteration:
            raise unexpected_end_of_stream_error(stream.char_position)

        raise unexpected_symbol_error(char, stream.char_position)
```

File: aidial_assistant/json_stream/json_node.py [hunk 3]
```diff
@@ -5,6 +5,9 @@
 from typing_extensions import override
 
 from aidial_assistant.json_stream.characterstream import CharacterStream
+from aidial_assistant.json_stream.exceptions import (
+    unexpected_end_of_stream_error,
+)
 
 
 class NodeResolver(ABC):
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_node.py]: Simply `NodeParser`, because it parses json. What does it resolve?

File: aidial_assistant/json_stream/json_array.py [hunk 4]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
-                value = await dependency_resolver.resolve(stream)
-                await self.listener.put(value)
-                if isinstance(value, ComplexNode):
-                    await value.parse(stream, dependency_resolver)
-                separate = True
-
-        await self.listener.put(None)
+    @staticmethod
+    async def read(
+        stream: CharacterStream, dependency_resolver: NodeResolver
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: `dependency_resolver` sounds really strange, what are those dependencies? The notion of 'dependency' is nowhere else to be found in the code. Simply `parser` is telling enough.

File: aidial_assistant/json_stream/json_array.py [hunk 5]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
     def token() -> str:
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: Comment for `json_bool.TRUE_STRING` and `json_bool.FALSE_STRING` - they are not unused anywhere.

File: aidial_assistant/json_stream/json_array.py [hunk 6]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
-                    await value.parse(stream, dependency_resolver)
-                separate = True
-
-        await self.listener.put(None)
+    @staticmethod
+    async def read(
+        stream: CharacterStream, dependency_resolver: NodeResolver
+    ) -> AsyncIterator[JsonNode]:
+        try:
+            normalised_stream = JsonNormalizer(stream)
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: Since you've chosen the american english spelling for `normalizer`, let's follow it consistently throughout the code: `normalized_stream`.

File: aidial_assistant/json_stream/json_parser.py [hunk 7]
```diff
@@ -38,34 +46,43 @@ def string_node(node: JsonNode) -> JsonString:
     return node
 
 
-class JsonParser:
-    @staticmethod
-    @asynccontextmanager
-    async def parse(stream: CharacterStream) -> AsyncGenerator[JsonNode, Any]:
-        root = JsonRoot()
+class RootNodeResolver(NodeResolver):
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_parser.py]: It's courious that in the module called `json_parser` the main classes aren't called like parsers. I suggest:

`RootNodeResolver` -> `JsonParser` (because it has nothing to do with the _rootness_ of the node, it could parse Json nodes of any nesting level. Instead it parses all kinds of Json nodes, not just one: `JsonString`, `JsonNumber` etc. So you could say it's `JsonAnyParser` or simply `JsonParser`)

`NodeResolver` -> `NodeParser`

`NodeResolver.resolve` -> `NodeParser.parse`

File: aidial_assistant/json_stream/json_parser.py [hunk 8]
```diff
@@ -38,34 +46,43 @@ def string_node(node: JsonNode) -> JsonString:
...
         try:
-            async with TaskGroup() as tg:
-                task = tg.create_task(JsonParser._parse_root(root, stream))
-                try:
-                    yield await root.node()
-                finally:
-                    await task
-        except ExceptionGroup as e:
-            raise e.exceptions[0]
+            normalised_stream = JsonNormalizer(stream)
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_parser.py]: `normalized_stream`

File: aidial_assistant/json_stream/json_parser.py [hunk 9]
```diff
@@ -38,34 +46,43 @@ def string_node(node: JsonNode) -> JsonString:
...
-            async for _ in stream:
-                pass
+            raise unexpected_end_of_stream_error(stream.char_position)
+
+        raise unexpected_symbol_error(char, stream.char_position)
+
+
+class JsonParser:
+    @staticmethod
+    async def parse(stream: CharacterStream) -> JsonNode:
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_parser.py]: It could be a simple method `parse_json`, no need to place it in a class.

File: aidial_assistant/json_stream/json_node.py [hunk 10]
```diff
@@ -33,22 +37,38 @@ def char_position(self) -> int:
         return self._char_position
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_node.py]: You could rename it to simply `pos` without loosing any information.

`node.pos` tells clearly that it's a position of a node in the source text.

`node.char_position` is too verbose, what kind of position could node have? Position in the JSON AST is rather "path" than "position".

File: aidial_assistant/json_stream/json_string.py [hunk 11]
```diff
@@ -93,7 +91,3 @@ async def escape(stream: CharacterStream) -> str:
             # Ignore when model cannot escape text properly
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_string.py]: Why ignore? The reference to the "model" here look strange. The json parser shouldn't know anything about the producer of json. Maybe configure it via flag "relaxed_char_escaping"?

File: aidial_assistant/json_stream/json_string.py [hunk 12]
```diff
@@ -93,7 +91,3 @@ async def escape(stream: CharacterStream) -> str:
...
             return char
             # raise ValueError(f"Unexpected escape sequence: \\{char}" + " at " + str(stream.char_position - 1))
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_string.py]: Dead code in comment

File: aidial_assistant/json_stream/json_string.py [hunk 13]
```diff
@@ -50,29 +35,42 @@ async def to_string_chunks(self) -> AsyncIterator[str]:
...
 
         if result:
             yield result
 
     @staticmethod
-    async def escape(stream: CharacterStream) -> str:
+    async def _escape(stream: CharacterStream) -> str:
         char = await anext(stream)
         if char == "u":
             unicode_sequence = "".join([await anext(stream) for _ in range(4)])  # type: ignore
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_string.py:73]: redundant square brackets, join takes iterable: `"".join(await anext(stream) for _ in range(4))`

Comment (adubovik) [inline aidial_assistant/json_stream/json_string.py:73]: `return chr(int(unicode_sequence, 16))`

Add a test with unicode symbol:

```python
    """
    {
      "str": "\\uFFFF",
    }
    """
```

File: aidial_assistant/json_stream/exceptions.py [hunk 14]
```diff
@@ -8,3 +8,9 @@ def unexpected_symbol_error(
     return JsonParsingException(
         f"Failed to parse json string: unexpected symbol {char} at position {char_position}"
```
Comment (adubovik) [inline aidial_assistant/json_stream/exceptions.py]: `json_bool` and `json_number` throw std JSON parser errors, because they use `json.loads` without any exception hanling.

`json_null` throws `unexpected_symbol_error`.

Let's use the same approach to parsing error messages everywhere.

Maybe catch the exception wherever you use `json.loads` and cast it to the library error messages, like `unexpected_token_error` (similar to `unexpected_symbol_error`).

File: aidial_assistant/json_stream/json_array.py [hunk 15]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
+            char = await anext(normalised_stream)
+            if not char == JsonArray.token():
+                raise unexpected_symbol_error(char, stream.char_position)
+
+            separate = False
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: `is_comma_expected` would be a more telling name

File: aidial_assistant/json_stream/json_array.py [hunk 16]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
-                separate = True
-
-        await self.listener.put(None)
+    @staticmethod
+    async def read(
+        stream: CharacterStream, dependency_resolver: NodeResolver
+    ) -> AsyncIterator[JsonNode]:
+        try:
+            normalised_stream = JsonNormalizer(stream)
+            char = await anext(normalised_stream)
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: 1. `JsonNormalizer` is a misnomer. I has nothing to do with JSON. It simply skips the whitespace *everywhere* in the stream.
2. It's not very obvious that `normalized_stream` also moves the pointer in the original `stream`.

It would be clearer to use a helper: `stream.skip_whitespaces()` anywhere you need it. Make it return the stream itself to allow chaining.

File: aidial_assistant/json_stream/json_array.py [hunk 17]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
+                    yield value
+
+                    if isinstance(value, ReadableNode):
+                        await value.read_to_end()
+                    separate = True
+        except StopAsyncIteration:
+            raise unexpected_end_of_stream_error(stream.char_position)
 
     @override
     async def to_string_chunks(self) -> AsyncIterator[str]:
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: Maybe rename `separate` to `is_first_element` (with inverted meaning).

File: aidial_assistant/json_stream/json_array.py [hunk 18]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
+                    value = await dependency_resolver.resolve(stream)
+                    yield value
+
+                    if isinstance(value, ReadableNode):
+                        await value.read_to_end()
+                    separate = True
+        except StopAsyncIteration:
+            raise unexpected_end_of_stream_error(stream.char_position)
 
     @override
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py:66]: Maybe rename `to_string_chunks` to `to_stream`?

File: aidial_assistant/json_stream/json_array.py [hunk 19]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
+            normalised_stream = JsonNormalizer(stream)
+            char = await anext(normalised_stream)
+            if not char == JsonArray.token():
+                raise unexpected_symbol_error(char, stream.char_position)
+
+            separate = False
+            while True:
+                char = await normalised_stream.apeek()
+                if char == "]":
+                    await anext(normalised_stream)
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: `await normalised_stream.askip()`

File: aidial_assistant/json_stream/json_array.py [hunk 20]
```diff
@@ -28,50 +29,40 @@ def type(self) -> str:
...
+                    break
+
+                if char == ",":
+                    if not separate:
+                        raise unexpected_symbol_error(
+                            char, stream.char_position
+                        )
+
+                    await anext(normalised_stream)
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_array.py]: `await normalised_stream.askip()`

File: aidial_assistant/json_stream/json_node.py [hunk 21]
```diff
@@ -33,22 +37,38 @@ def char_position(self) -> int:
...
+    def _accumulate(self, element: TElement):
         pass
 
+    async def read_to_end(self):
+        async for _ in self:
+            pass
+
 
-class PrimitiveNode(JsonNode[T], ABC, Generic[T]):
+class PrimitiveNode(JsonNode[TValue], ABC, Generic[TValue]):
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_node.py]: From the names of the classes it's not clear that they are mitually exclusive.
I suggest to rename:

`PritimiveNode` -> `AtomicNode`

`ReadableNode` -> `CompoundNode`

File: aidial_assistant/json_stream/json_object.py [hunk 22]
```diff
@@ -55,47 +44,55 @@ async def get(self, key: str) -> JsonNode:
...
+                    if isinstance(value, ReadableNode):
+                        await value.read_to_end()
+                    separate = True
+                else:
+                    raise unexpected_symbol_error(char, stream.char_position)
+        except StopAsyncIteration:
+            raise unexpected_end_of_stream_error(stream.char_position)
 
     @override
     async def to_string_chunks(self) -> AsyncIterator[str]:
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_object.py]: `separate` -> `is_first_entry` with inverted meaning

File: aidial_assistant/json_stream/json_object.py [hunk 23]
```diff
@@ -55,47 +44,55 @@ async def get(self, key: str) -> JsonNode:
...
+                    colon = await anext(normalised_stream)
+                    if not colon == ":":
+                        raise unexpected_symbol_error(
+                            colon, stream.char_position
+                        )
+
+                    value = await dependency_resolver.resolve(stream)
+                    yield key, value
+
+                    if isinstance(value, ReadableNode):
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_object.py]: You could move `read_to_end` for `JsonObject/Array/String` to `RootNodeResolver.resolve` to place them in the same method as `PrimitiveNode.collect(stream)` calls, which do the conceptually similar thing - move the stream pointer behind the entity which is being parsed.

File: aidial_assistant/json_stream/json_node.py [hunk 24]
```diff
@@ -59,11 +79,14 @@ async def to_string_chunks(self) -> AsyncIterator[str]:
 
     @staticmethod
     async def collect(stream: CharacterStream) -> str:
```
Comment (adubovik) [inline aidial_assistant/json_stream/json_node.py]: `PrimitiveNode.collect` is an analog of `ReadableNode.read_to_end`. Maybe give it the same name then: `PrimitiveNode.read_to_end`.

File: aidial_assistant/json_stream/exceptions.py [hunk 25]
```diff
@@ -1,10 +1,21 @@
...
+
+
+def unexpected_end_of_stream_error(char_position: int) -> JsonParsingException:
+    return JsonParsingException("Unexpected end of stream.", char_position)
+
+
+def invalid_sequence_error(
+    sequence: str, char_position: int
+) -> JsonParsingException:
+    return JsonParsingException(f"Invalid sequence {sequence}.", char_position)
```
Comment (adubovik) [inline aidial_assistant/json_stream/exceptions.py]: What does `sequence` mean? Any string is a sequence of character, so it doesn't help at all.

It think the error message `Can't parse {expected_type} from the string '{string}'.` is much more comprehensive.

File: aidial_assistant/json_stream/exceptions.py [hunk 26]
```diff
@@ -1,10 +1,21 @@
...
+        )
 
 
 def unexpected_symbol_error(
     char: str, char_position: int
 ) -> JsonParsingException:
-    return JsonParsingException(
-        f"Failed to parse json string: unexpected symbol {char} at position {char_position}"
-    )
+    return JsonParsingException(f"Unexpected symbol {char}.", char_position)
```
Comment (adubovik) [inline aidial_assistant/json_stream/exceptions.py]: Maybe quote the symbol to make absolutely clear what it is.

`Unexpected symbol '{char}'.`

**Review by adubovik — APPROVED** [review]
